File: MB-Whatsapp/memory_py/semantic_memory_service.py
from __future__ import annotations

import os
from dataclasses import replace
from typing import Any, Optional

from .audit import InMemoryAuditSink, build_memory_audit_event, build_memory_audit_hash
from .jsonl_adapter import JsonlMemoryAdapter
from .retriever import MemoryRetriever
from .sanitizer import MemorySanitizer
from .types import MemoryEntry, MemoryFeatureFlags, MemoryScope


def _read_bool(value: Optional[str], fallback: bool) -> bool:
    if value is None:
        return fallback
    normalized = value.strip().lower()
    if normalized in {"1", "true", "yes", "on"}:
        return True
    if normalized in {"0", "false", "no", "off"}:
        return False
    return fallback


def load_memory_feature_flags(env: dict[str, str] | None = None) -> MemoryFeatureFlags:
    values = env if env is not None else os.environ
    return MemoryFeatureFlags(
        enabled=_read_bool(values.get("SEMANTIC_MEMORY_ENABLED"), False),
        shadow_mode=_read_bool(values.get("SEMANTIC_MEMORY_SHADOW_MODE"), True),
        vector_enabled=_read_bool(values.get("SEMANTIC_MEMORY_VECTOR_ENABLED"), False),
        write_enabled=_read_bool(values.get("SEMANTIC_MEMORY_WRITE_ENABLED"), False),
        patient_scope_enabled=_read_bool(values.get("SEMANTIC_MEMORY_PATIENT_SCOPE_ENABLED"), False),
    )


class SemanticMemoryService:
    def __init__(
        self,
        backend: JsonlMemoryAdapter,
        flags: MemoryFeatureFlags | None = None,
        sanitizer: MemorySanitizer | None = None,
        audit_sink: InMemoryAuditSink | None = None,
    ) -> None:
        self.backend = backend
        self.flags = flags or load_memory_feature_flags()
        self.sanitizer = sanitizer or MemorySanitizer()
        self.retriever = MemoryRetriever(backend)
        self.audit_sink = audit_sink

    def remember(self, entry: MemoryEntry) -> dict[str, Any]:
        scope = MemoryScope(
            scope=entry.scope,
            tenant_id=entry.tenant_id,
            doctor_id=entry.doctor_id,
            patient_id=entry.patient_id,
        )
        try:
            if not self.flags.enabled:
                self._audit("remember", entry.trace_id, scope, False, True, "semantic_memory_disabled")
                return {"stored": False, "shadowed": False, "fallback_used": True, "reason": "semantic_memory_disabled"}

            if entry.patient_id and not self.flags.patient_scope_enabled:
                self._audit("remember", entry.trace_id, scope, False, True, "patient_scope_disabled")
                return {"stored": False, "shadowed": False, "fallback_used": True, "reason": "patient_scope_disabled"}

            sanitized = self.sanitizer.sanitize_content(entry.content)
            if sanitized.get("blocked"):
                reason = sanitized.get("reason") or "sanitizer_blocked"
                self._audit("remember", entry.trace_id, scope, False, True, reason)
                return {"stored": False, "shadowed": False, "fallback_used": True, "reason": reason}

            sanitized_metadata = self.sanitizer.sanitize_metadata(entry.metadata)
            sanitized_metadata["sanitizer_redactions"] = sanitized.get("redactions", [])
            prepared = replace(
                entry,
                sanitized_content=sanitized["sanitized_content"],
                metadata=sanitized_metadata,
            )
            prepared.audit_hash = build_memory_audit_hash(prepared)

            if self.flags.shadow_mode or not self.flags.write_enabled:
                self._audit("remember", prepared.trace_id, scope, True, False, "shadow_mode_write_suppressed")
                return {
                    "stored": False,
                    "shadowed": True,
                    "entry": prepared,
                    "fallback_used": False,
                    "reason": "shadow_mode_write_suppressed",
                }

            stored = self.backend.append(prepared)
            self._audit("remember", stored.trace_id, scope, True, False)
            return {"stored": True, "shadowed": False, "entry": stored, "fallback_used": False}
        except Exception:
            self._audit("fallback", entry.trace_id, scope, False, True, "semantic_memory_remember_failed")
            return {"stored": False, "shadowed": False, "fallback_used": True, "reason": "semantic_memory_remember_failed"}

    def recall(self, query: str, scope: MemoryScope, trace_id: str, limit: int = 20) -> dict[str, Any]:
        try:
            if not self.flags.enabled:
                self._audit("recall", trace_id, scope, True, True, "semantic_memory_disabled")
                return {
                    "entries": [],
                    "fallback_used": True,
                    "reason": "semantic_memory_disabled",
                    "retrieval_mode": "disabled",
                }

            if scope.patient_id and not self.flags.patient_scope_enabled:
                self._audit("recall", trace_id, scope, True, True, "patient_scope_disabled")
                return {
                    "entries": [],
                    "fallback_used": True,
                    "reason": "patient_scope_disabled",
                    "retrieval_mode": "disabled",
                }

            result = self.retriever.search(query=query, scope=scope, limit=limit)
            self._audit("recall", trace_id, scope, True, False)
            return {**result, "fallback_used": False}
        except Exception:
            self._audit("fallback", trace_id, scope, False, True, "semantic_memory_recall_failed")
            return {
                "entries": [],
                "fallback_used": True,
                "reason": "semantic_memory_recall_failed",
                "retrieval_mode": "disabled",
            }

    def _audit(
        self,
        action: str,
        trace_id: str,
        scope: MemoryScope,
        success: bool,
        fallback_used: bool,
        reason: str | None = None,
    ) -> None:
        if not self.audit_sink:
            return
        self.audit_sink.record(
            build_memory_audit_event(
                action=action,
                backend="jsonl",
                trace_id=trace_id,
                scope=scope,
                success=success,
                fallback_used=fallback_used,
                reason=reason,
            )
        )

File: MB-Chat/memory/semantic-memory-service.ts
import { randomUUID } from "node:crypto";
import { loadMemoryFeatureFlags } from "./feature-flags";
import { buildMemoryAuditEvent, buildMemoryAuditHash, type MemoryAuditSink } from "./memory-audit";
import { MemoryRetriever } from "./memory-retriever";
import { MemorySanitizer, defaultMemorySanitizer } from "./memory-sanitizer";
import type {
  MemoryBackend,
  MemoryFeatureFlags,
  RecallInput,
  RecallResult,
  RememberInput,
  RememberResult,
} from "./types";

export type SemanticMemoryServiceOptions = {
  backend: MemoryBackend;
  flags?: MemoryFeatureFlags;
  sanitizer?: MemorySanitizer;
  auditSink?: MemoryAuditSink;
};

export class SemanticMemoryService {
  private readonly backend: MemoryBackend;
  private readonly flags: MemoryFeatureFlags;
  private readonly sanitizer: MemorySanitizer;
  private readonly retriever: MemoryRetriever;
  private readonly auditSink?: MemoryAuditSink;

  constructor(options: SemanticMemoryServiceOptions) {
    this.backend = options.backend;
    this.flags = options.flags ?? loadMemoryFeatureFlags();
    this.sanitizer = options.sanitizer ?? defaultMemorySanitizer;
    this.retriever = new MemoryRetriever(this.backend);
    this.auditSink = options.auditSink;
  }

  async remember(input: RememberInput): Promise<RememberResult> {
    const entryScope = {
      scope: input.entry.scope,
      tenant_id: input.entry.tenant_id,
      doctor_id: input.entry.doctor_id,
      patient_id: input.entry.patient_id,
    };

    try {
      if (!this.flags.enabled) {
        await this.audit("remember", input.entry.trace_id, false, true, "semantic_memory_disabled", entryScope);
        return { stored: false, shadowed: false, fallback_used: true, reason: "semantic_memory_disabled" };
      }

      if (input.entry.patient_id && !this.flags.patientScopeEnabled) {
        await this.audit("remember", input.entry.trace_id, false, true, "patient_scope_disabled", entryScope);
        return { stored: false, shadowed: false, fallback_used: true, reason: "patient_scope_disabled" };
      }

      const sanitized = this.sanitizer.sanitizeContent(input.entry.content);
      if (sanitized.blocked) {
        await this.audit("remember", input.entry.trace_id, false, true, sanitized.reason ?? "sanitizer_blocked", entryScope);
        return { stored: false, shadowed: false, fallback_used: true, reason: sanitized.reason };
      }

      const entry = {
        ...input.entry,
        id: input.entry.id || randomUUID(),
        sanitized_content: sanitized.sanitized_content,
        metadata: {
          ...this.sanitizer.sanitizeMetadata(input.entry.metadata),
          sanitizer_redactions: sanitized.redactions,
        },
      };
      entry.audit_hash = buildMemoryAuditHash(entry);

      if (this.flags.shadowMode || !this.flags.writeEnabled) {
        await this.audit("remember", entry.trace_id, true, false, "shadow_mode_write_suppressed", entryScope);
        return { stored: false, shadowed: true, entry, fallback_used: false, reason: "shadow_mode_write_suppressed" };
      }

      const stored = await this.backend.append(entry);
      await this.audit("remember", stored.trace_id, true, false, undefined, entryScope);
      return { stored: true, shadowed: false, entry: stored, fallback_used: false };
    } catch (error) {
      await this.audit(
        "fallback",
        input.entry.trace_id,
        false,
        true,
        error instanceof Error ? error.message : String(error),
        entryScope,
      );
      return { stored: false, shadowed: false, fallback_used: true, reason: "semantic_memory_remember_failed" };
    }
  }

  async recall(input: RecallInput): Promise<RecallResult> {
    try {
      if (!this.flags.enabled) {
        await this.audit("recall", input.trace_id, true, true, "semantic_memory_disabled", input.scope);
        return {
          entries: [],
          fallback_used: true,
          reason: "semantic_memory_disabled",
          retrieval_mode: "disabled",
        };
      }

      if (input.scope.patient_id && !this.flags.patientScopeEnabled) {
        await this.audit("recall", input.trace_id, true, true, "patient_scope_disabled", input.scope);
        return {
          entries: [],
          fallback_used: true,
          reason: "patient_scope_disabled",
          retrieval_mode: "disabled",
        };
      }

      const result = await this.retriever.search({
        query: input.query,
        scope: input.scope,
        filters: input.filters,
      });
      await this.audit("recall", input.trace_id, true, false, undefined, input.scope);
      return {
        entries: result.entries,
        fallback_used: false,
        retrieval_mode: result.retrieval_mode,
      };
    } catch (error) {
      await this.audit(
        "fallback",
        input.trace_id,
        false,
        true,
        error instanceof Error ? error.message : String(error),
        input.scope,
      );
      return {
        entries: [],
        fallback_used: true,
        reason: "semantic_memory_recall_failed",
        retrieval_mode: "disabled",
      };
    }
  }

  async trainFromMedicalChat(input: MedicalChatLearningRecord): Promise<boolean> {
    try {
      const entry = {
        id: input.id,
        tenant_id: "medical-chat",
        doctor_id: input.role === "DOCTOR" ? input.role : "system",
        patient_id: input.sessionId,
        scope: "session",
        kind: "training-record",
        content: input.queryHash,
        sanitized_content: input.explicitTeaching?.sanitizedText || "",
        source: "medical-chat-learning",
        confidence: input.decision.confidence,
        tags: input.concepts,
        created_at: input.recordedAt,
        metadata: {
          modality: input.modality,
          citations: input.citationCount,
          outcome: input.outcome,
          decision: input.decision,
        },
      };

      await this.backend.append(entry);
      return true;
    } catch (error) {
      console.error("Failed to train from medical chat record:", error);
      return false;
    }
  }

  private async audit(
    action: "remember" | "recall" | "fallback",
    traceId: string,
    success: boolean,
    fallbackUsed: boolean,
    reason?: string,
    scope?: RecallInput["scope"],
  ): Promise<void> {
    if (!this.auditSink) {
      return;
    }

    await this.auditSink.record(
      buildMemoryAuditEvent({
        action,
        backend: "jsonl",
        trace_id: traceId,
        scope,
        success,
        fallback_used: fallbackUsed,
        reason,
      }),
    );
  }
}
